src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -0,0 +1,1146 @@
+            }
+
+            lock (_gate)
+            {
+                if (!ContainsMetadataReference(fullPath, properties))
+                {
+                    throw new InvalidOperationException("The metadata reference does not exist in this project.");
+                }
+
+                _allMetadataReferences.MultiRemove(fullPath, properties);
+
+                if (_activeBatchScopes > 0)
+                {
+                    if (!_metadataReferencesAddedInBatch.Remove((fullPath, properties)))
+                    {
+                        _metadataReferencesRemovedInBatch.Add((fullPath, properties));
+                    }
+                }
+                else
+                {
+                    _workspace.ApplyChangeToWorkspace(w =>
+                    {
+                        var projectReference = _workspace.TryRemoveConvertedProjectReference(Id, fullPath, properties);
+
+                        // If this was converted to a project reference, we have now recorded the removal -- let's remove it here too
+                        if (projectReference != null)
+                        {
+                            w.OnProjectReferenceRemoved(Id, projectReference);
+                        }
+                        else
+                        {
+                            // TODO: find a cleaner way to fetch this
+                            var metadataReference = w.CurrentSolution.GetProject(Id).MetadataReferences.Cast<PortableExecutableReference>()
+                                                                                    .Single(m => m.FilePath == fullPath && m.Properties == properties);
+
+                            w.OnMetadataReferenceRemoved(Id, metadataReference);
+                        }
+                    });
+                }
+            }
+        }
+
+        #endregion
+
+        #region Project Reference Addition/Removal
+
+        public void AddProjectReference(ProjectReference projectReference)
+        {
+            if (projectReference == null)
+            {
+                throw new ArgumentNullException(nameof(projectReference));
+            }
+
+            lock (_gate)
+            {
+                if (ContainsProjectReference(projectReference))
+                {
+                    throw new ArgumentException("The project reference has already been added to the project.");
+                }
+
+                if (_activeBatchScopes > 0)
+                {
+                    if (!_projectReferencesRemovedInBatch.Remove(projectReference))
+                    {
+                        _projectReferencesAddedInBatch.Add(projectReference);
+                    }
+                }
+                else
+                {
+                    _workspace.ApplyChangeToWorkspace(w => w.OnProjectReferenceAdded(Id, projectReference));
+                }
+            }
+        }
+
+        public bool ContainsProjectReference(ProjectReference projectReference)
```

Comment:
>ContainsProjectReference [](start = 20, length = 24)

Should we have -NoLock version of this, so that we don't need to nest locks?

Reply:
I'm going to go with _maybe_ right now. I don't want to double the method count here for all these helpers in both Lock and _NoLock variants, so we'll see.

There's a separate item though for switching this over to a SemaphoreSlim so we can have better async locking anyways. ([item 1](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/677917) and [item 2](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/677921).

Not going to resolve this but am not going to address it unless we run out of higher priority items.